### PR TITLE
[CPU] Enable cpu_convert utility function with AVX_512 FP16

### LINKED ISSUE #21809: [Good First Issue]: enable cpu_convert utility function with AVX_512 FP16 store/load instructions
### Context

OpenVINO CPU Plugin internally uses a utility function cpu_convert. In the present moment this function uses AVX2 instruction to load/store fp16 data. Probably it is more efficient to use AVX512 fp16 instruction for platforms with AVX512 ISA.

https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp#L29-L142 

### What needs to be done?

1. augment cpu_convert function by adding AVX512 fp16 load/store instruction support.
2. add unit test for the feature
3. benchmark the performance of AVX2 and AVX512 implementation with different sizes of workload.
4. based on the benchmarking result to decide if AVX512 implementation is better for performance improvement.  

### Example Pull Requests

_No response_

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)

### HW Preparation
Before you make decision to take the contribution, please make sure you have the AVX512 CPU platform in hand for development and performance debugging and profiling.
https://www.intel.com/content/www/us/en/architecture-and-technology/avx-512-overview.html

### Contact points

@ceciliapeng2011 @yuxu42 @luo-cheng2021 

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by siddhant-0707**
### Details:
 - Augmented cpu_convert function by adding AVX512 fp16 load/store instruction support.
 - adding tests

### Tickets:
 - Closes #21809

### CODE REVIEW

**Comment by siddhant-0707** [comment]
Hey @ceciliapeng2011 could you guide me a little bit on where and how to create the required tests for this?

**Comment by siddhant-0707** [comment]
Hey @ceciliapeng2011, please review, thanks!

**Comment by ceciliapeng2011** [comment]
thanks for contribution @siddhant-0707 
For unit test, maybe you could refer to this unit test - https://github.com/openvinotoolkit/openvino/pull/22174/files#diff-68058ac8f7dc6ca18caffd8e7ff762035dcd7ee15f5ccf4b78856a9a50639adf

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp [hunk 1]
```diff
@@ -125,10 +150,15 @@ class jit_convert_array : public jit_kernel {
...
+            auto& generator = static_cast<jit_generator&>(converter);
+            generator.create_kernel();
+            return (fn_t)generator.jit_ker();
+        } else if (mayiuse(cpu_isa_t::avx2) && dnnl::impl::cpu::x64::cpu().has(Xbyak::util::Cpu::tF16C)) {
             static jit_convert_array converter(convert_vec<src_t, dst_t>, sizeof(src_t), sizeof(dst_t));
-            auto & generator = static_cast<jit_generator&>(converter);
+            auto& generator = static_cast<jit_generator&>(converter);
             generator.create_kernel();
             return (fn_t)generator.jit_ker();
         }
```
Comment (ceciliapeng2011) [inline src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:164]: Looks the if and else branches are using the same code piece? How do you differentiate which specialized 
template is used?

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp [hunk 2]
```diff
@@ -55,6 +55,31 @@ void convert_vec<float, ov::float16>(jit_generator & gen,
     gen.movdqu(gen.xword[dst], f16vec);
 }
 
+// AVX512 instructions for ov::float16 to float conversion
+template <>
+void convert_vec<ov::float16, float>(jit_generator & gen,
+                                        const RegExp & src,
+                                        const RegExp & dst) {
```
Comment (ceciliapeng2011) [inline src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:62]: How do you differentiate which specialized template for avx2 and avx512?

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp [hunk 3]
```diff
@@ -55,6 +55,31 @@ void convert_vec<float, ov::float16>(jit_generator & gen,
...
+    auto const & f16vec = gen.zmm3;
+    auto const & f32vec = gen.zmm4;
+
+    gen.vmovups(f32vec, gen.zword[src]);
+    gen.vcvtps2ph(f16vec, f32vec, 0);
+    gen.vmovdqu16(gen.zword[dst], f16vec);
+}
+
```
Comment (ceciliapeng2011) [inline src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:82]: When you do performance comparison between with AVX2 and AVX512, maybe you could change below lines to match vector size to adapt to avx512.

```
constexpr size_t vlen = 8u;
constexpr size_t vlen_log2 = 3;
```

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp (hunk 1 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:169]: Probably we can use ISA as the template parameter here.

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp [hunk 4]
```diff
@@ -27,12 +27,17 @@ using namespace dnnl::impl::cpu::x64;
 using namespace Xbyak;
 
 template <typename src_t, typename dst_t>
-void convert_vec(jit_generator & gen,
+void convert_vec_avx2(jit_generator & gen,
+                 const RegExp & src,
+                 const RegExp & dst);
+
+template <typename src_t, typename dst_t>
+void convert_vec_avx512(jit_generator & gen,
```
Comment (ceciliapeng2011) [inline src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:35]: Probably we can use ISA as the template parameter here. Refer to https://github.com/ceciliapeng2011/openvino/blob/cecilia/kvcache_fp16/src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp (hunk 3 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:87]: Do you have a unit test for accuracy and performance check please?

**Comment by siddhant-0707** [comment]
Hey, on my machine I am able to run the test (it uses `avx2`)

```
// Using AVX2
[ RUN      ] cpu_convert.AVX512_fp16_load_store
size 1000: 44 microseconds
size 10000: 4 microseconds
size 100000: 33 microseconds
size 1000000: 596 microseconds
[       OK ] cpu_convert.AVX512_fp16_load_store (5 ms)
[----------] 1 test from cpu_convert (5 ms total)
```

**Comment by ceciliapeng2011** [comment]
CI probably cannot benchmark the performance across platforms. Do you have a local machine with AVX512?

File: src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp (hunk 3 above)
Comment (ceciliapeng2011) [reply src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp:82]: > When you do performance comparison between with AVX2 and AVX512, maybe you could change below lines to match vector size to adapt to avx512.
> 
> ```
> constexpr size_t vlen = 8u;
> constexpr size_t vlen_log2 = 3;
> ```

Could you please try this to see if any benefits?

**Comment by siddhant-0707** [comment]
No, unfortunately I don't

**Comment by siddhant-0707** [comment]
Hey @ceciliapeng2011 finally got a machine with AVX512 capability.
Here are the results I collected after changing the lines you indicated to:
```
constexpr size_t vlen = 16u;
constexpr size_t vlen_log2 = 4;
```

```
[ RUN      ] cpu_convert.AVX512_fp16_load_store
size 1000: 645 microseconds
size 10000: 1055 microseconds
size 100000: 48 microseconds
size 1000000: 122 microseconds
size 10000000: 1398 microseconds
[       OK ] cpu_convert.AVX512_fp16_load_store (17 ms)


[ RUN      ] cpu_convert.AVX512_fp16_load_store
size 1000: 674 microseconds
size 10000: 1112 microseconds
size 100000: 41 microseconds
size 1000000: 143 microseconds
size 10000000: 1293 microseconds
[       OK ] cpu_convert.AVX512_fp16_load_store (17 ms)
```

**Comment by ceciliapeng2011** [comment]
> Hey @ceciliapeng2011 finally got a machine with AVX512 capability. Here are the results I collected after changing the lines you indicated to:
> 
> ```
> constexpr size_t vlen = 16u;
> constexpr size_t vlen_log2 = 4;
> ```
> 
> ```
> [ RUN      ] cpu_convert.AVX512_fp16_load_store
> size 1000: 645 microseconds
> size 10000: 1055 microseconds
> size 100000: 48 microseconds
> size 1000000: 122 microseconds
> size 10000000: 1398 microseconds
> [       OK ] cpu_convert.AVX512_fp16_load_store (17 ms)
> 
> 
> [ RUN      ] cpu_convert.AVX512_fp16_load_store
> size 1000: 674 microseconds
> size 10000: 1112 **microseconds**
> size 100000: 41 microseconds
> size 1000000: 143 microseconds
> size 10000000: 1293 microseconds
> [       OK ] cpu_convert.AVX512_fp16_load_store (17 ms)
> ```

Glad you have the AVX512 machine and continue the job! Great!
So would you please benchmark the workload with both AVX2 and AVX512 on the same machine?

please make sure the scaling governors mode of your machine is **performance** (default is **powersave**) before benchmarking.
You could set it with Linux command -
`echo "performance " | sudo tee /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor`

File: src/plugins/intel_cpu/tests/unit/nodes/avx_512_test.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    std::vector<size_t> sizes = {1000, 10000, 100000, 1000000};
+    for (size_t size : sizes) {
+        std::vector<float> input_data(size, 1.23f);
+        std::vector<short> output_data(size);
+
+        auto start = high_resolution_clock::now();
+        cpu_convert(input_data.data(), output_data.data(), ov::element::f32, ov::element::f16, size);
+        auto stop = high_resolution_clock::now();
+        auto duration = duration_cast<microseconds>(stop - start);
+        std::cout << "size " << size << ": " << duration.count() << " microseconds\n";
```
Comment (ceciliapeng2011) [inline src/plugins/intel_cpu/tests/unit/nodes/avx_512_test.cpp:25]: We'd better check the result to see if it is as expected.

**Comment by ceciliapeng2011** [comment]
@siddhant-0707 From my perspective, this PR still needs the following two unit tests -
1. cross-compare the performance number of converting with avx2 and avx512 fp16 with different workloads;
2. validate the output result

**Comment by mlukasze** [comment]
hey @siddhant-0707, will you have a time to finish this PR?

**Pull request closed without merging**